xla/service/collective_quantizer.cc

```diff
@@ -50,14 +60,121 @@ auto ConvertToNarrowerType() {
+// division by a broadcasted scalar operating on a type conversion, or a plain
+// type conversion to a wider type. Unary bitcast, copy, reshape or slice ops
+// may follow the dequantization or type conversion.
+std::optional<ConversionSubgraph> IsSupportedDequantization(
+    HloInstruction* instr) {
+  ConversionSubgraph subgraph;
+  absl::flat_hash_set<int> visited_instrs;
+  std::vector<HloInstruction*> candidate_subgraph =
+      FindDequantizationSubgraphRecursive(instr, visited_instrs,
+                                          std::vector<HloInstruction*>{});
+  std::reverse(candidate_subgraph.begin(), candidate_subgraph.end());
+
+  // In the dequantization case, the type conversion is followed by a
+  // multiplication or division by a broadcasted scalar.
+  if (candidate_subgraph.size() > 1 &&
+      (Match(
+           candidate_subgraph[1],
+           m::MultiplyAnyOrder(&subgraph.binary, m::Convert(&subgraph.convert),
+                               ScalarBroadcast(&subgraph.scale_bcast))) ||
+       Match(candidate_subgraph[1],
+             m::Divide(&subgraph.binary, m::Convert(&subgraph.convert),
+                       ScalarBroadcast(&subgraph.scale_bcast))))) {
+    subgraph.unaries = {candidate_subgraph.begin() + 2,
+                        candidate_subgraph.end()};
+  } else if (candidate_subgraph.size() > 0 &&
+             Match(candidate_subgraph[0], m::Convert(&subgraph.convert))) {
+    subgraph.unaries = {candidate_subgraph.begin() + 1,
+                        candidate_subgraph.end()};
+  } else {
+    VLOG(5) << "Did not find type conversion or dequantization pattern.";
+    return std::nullopt;
+  }
+
+  // The collected unary ops between dequantization/type conversion and
+  // collective may only include bitcast, copy, reshape and slice instructions.
+  for (HloInstruction* unary : subgraph.unaries) {
+    if (!Match(unary, m::AnyOf<HloInstruction>(m::Bitcast(), m::Copy(),
+                                               m::Reshape(), m::Slice()))) {
+      VLOG(5) << "Unexpected instruction in unary ops.";
+      return std::nullopt;
+    }
+  }
+  return std::make_optional<ConversionSubgraph>(subgraph);
+}
+
 // Returns true iff instr describes a quantization, i.e. a multiplication or
 // division by a broadcasted scalar followed by a clamp and a type conversion,
 // or a plain type conversion to a narrower type. Unary bitcast, copy, reshape
 // or slice ops with one user may precede the quantization or type conversion.
-bool IsSupportedQuantization(HloInstruction* instr, HloInstruction** convert,
-                             HloInstruction** binary, HloInstruction** clamp,
-                             HloInstruction** scale_bcast,
-                             std::vector<HloInstruction*>& unary_ops) {
+std::optional<ConversionSubgraph> IsSupportedQuantization(
+    HloInstruction* instr) {
+  ConversionSubgraph subgraph;
   std::vector<HloInstruction*> ops;
   while (instr->user_count() <= 1) {
```

Comment:
Maybe it would be easier to match the convert by creating a `handleConvert` method and matching up through operands instead of down through users. That way its more consistent with with the DQ case, and you could share the `FindDequantizationSubgraphRecursive` method.